# BAM to count table notes (seminar 5)  

## SAM files  

* "Sequence Alignment Map"  
* document begins with headers (lines beginning with"@"); followed by alignments (individual entries for each alignment, with 11 mandatory fields each)  
* each read can align to multiple locations in the reference genome, but each will be reported as a separate alignment  
* mandatory alignment fields:  
    - QNAME: FASTQ name  
    - FLAG: integer value that encodes metadata about read type (e.g. paired-end, forward/reverse strand); see [this decoder tool](http://broadinstitute.github.io/picard/)  
    - RNAME: chromosome name/identifier  
    - POS: left-most position of alignment in the genome  
    - MAPQ: quality of mapping (in Phred-scaled probability value; higher = better)  
    - CIGAR: # of matching (M), insertion (I), or deletion (D) bases, in order  
    - RNEXT: for paired-end reads, the chromosome name/identifier of the mate pair (* = not paired-end data) 
    - PNEXT: position of mate pair in the genome  
    - TLEN: template length (paired-end only)  
    - SEQ: raw sequence as originally defined in the FASTQ file  
    - QUAL: Phred quality score of each base (all bases are simply set to 2, indicating a 0.01995 probability of incorrect call)  

## SAM file to count table: packages 

* Command-line tools: samtools (sorting, indexing), HTSeq (count table generation)  
    - to use these you must login via PuTTy (a unix terminal) & the course's bcgsc account  
* note: you must use a tool for count table generation that is matched with the alignment tool that generated the SAM files
    - e.g. HTSeq works for alignments generated by STAR

##  UNIX commands to know  

* general structure: utility flag argument
    - utility: what to do
    - flag: option/modification
    - argument: specific information (e.g. file name)
    
* ls: __list__ contents of wd (ls)
* pwd: __print working directory__ path (pwd)
* cd: __change__ directory (cd path_to_new_directory)
* less: __display contents of a file__ (less file_name)... scroll using arrow keys, press Q to exit
* cp: __copy__ a file in same directory file (cp old_name new_name)
* mv: __move__ a file to new directory (mv file_name new_directory_path)
* rm: __remove__ a file PERMANENTLY (rm file_name)
* be careful of sudo: __super user do__ is a higher-level command, that removes some of the protection against doing bad things, like deleting all files on your computer... do not use unless you know what you're doing, and trust the source where you're getting the script from
* for more, see [this tutorial](https://www.davidbaumgold.com/tutorials/command-line/)

## Workflow  

* _will add once done at tutorial_
  - _oy, not done at tutorial because they don't have server access for us_

* rough workflow:
  - sort and index using samtools
  - generate count table using htseq

## Next steps  

* once you have counts, you can use edgeR to normalize to library size (cpm), etc.
