---
title: "Assignment 1"
author: "Jennifer Grants"
date: "February 13, 2019"
output: 
  html_document:
    toc: true
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(knitr)
library(reshape2)
```

# Brief summary of study goal, design, data

* __Link:__ [https://www.ncbi.nlm.nih.gov/pubmed/25904789](https://www.ncbi.nlm.nih.gov/pubmed/25904789)
* __Goal:__ To identify genes highly expressed in inner ear hair cells (HCs) _vs._ surrounding cells (SCs) during development.
* __Design factors:__
    - 16 samples with a matrix design of 3 factors
        - cell types: HCs, SCs
        - tissue sources: cochlea, utricle
        - developmental stage: E16, P0, P4, P7
* __Gene expression analysis method in brief:__
    - mRNA amplification (NuGEN  Ovation 3â€²-DGE System)
    - Illumina RNA-seq

    
## Data pre-processing summary

* The raw reads have been mapped and processed into gene expression values using an established RNA-Seq pipeline at the Pavlidis Lab during data curation in Gemma. 
* Expression values are given in Counts per Million (CPM).
* Quantile normalization has been done as part of the data processing pipeline.


# Question 1: Data Inspection and Basic Manipulation
## Q1.1 Importing the data and getting familiar with it

### Read the datasets into RStudio
* Design matrix:   
```{r}
design <- readRDS("./data/gse60019_experiment_design.RDS")

kable(design)
dim(design)
```

* Expresion matrix:
```{r}
expr <- readRDS("./data/gse60019_expression_matrix.RDS")

kable(head(expr))

dim(expr)
```

### How many genes are there?
```{r}
ngene <- nrow(expr)
```

> There are `r ngene` genes in the expression matrix.

### How many samples are there?
```{r}
nsamp <- nrow(design)
```

> There are `r nsamp` samples in the experiment design.

### How many factors are in our experimental design? How may levels per factor? List out the levels for each factor.
```{r}
str(design)

op <- levels(design$organism_part)
ct <- levels(design$cell_type)
tp <- levels(design$time_point)
bt <- levels(design$batch)
```

> There are 4 factors in the experimental design (`r colnames(design[,2:5])`).

> Number of levels per factor:

  - `r colnames(design[,2])`: `r length(op)`  
  - `r colnames(design[,3])`: `r length(ct)`  
  - `r colnames(design[,4])`: `r length(tp)`  
  - `r colnames(design[,5])`: `r length(bt)`  
  
> Level names:  

  - `r colnames(design[,2])`: `r op`  
  - `r colnames(design[,3])`: `r ct`  
  - `r colnames(design[,4])`: `r tp`  
  - `r colnames(design[,5])`: `r bt`  
  
## Q1.2 Data manipulation
Here I will create a new design variable, `age`, which will be a numerical representation of the factor `time_point`. This assumes that P0 = 18 days.
```{r}
design_age <- mutate(design, age = time_point) 
levels(design_age$age) <- c(16, 18, 18+4, 18+7)

design_age$age <- as.numeric(as.character(design_age$age)) ## this is important to avoid loss of numerical information

# check
str(design_age)
```


## Q1.3 Single gene graphing
Here I will graph expression of Vegfa vs. age, with regression lines for the 2 cell types.
```{r warning=FALSE, message=FALSE}
# take Vegfa row of expression matrix and melt
vegfa <- filter(expr, gene == "Vegfa") %>%
  column_to_rownames(var = "gene") %>%
  melt()

colnames(vegfa) <- c("sample", "expression")

# join with metadata
vegfa <- left_join(vegfa, design_age, by = "sample")

# graph age vs. expression of gene
ggplot(vegfa, aes(age, expression, colour = cell_type)) +
  geom_jitter(width = 0.1) +
  geom_smooth(method = "glm", se = F) +
  ggtitle("Vegfa") +
  labs(y = "expression (cpm)")
```

### Is there sign of interaction between cell_type and age for Vegfa?

> There appears to be no interaction between cell type and age for Vegfa, as the lines are parallel between the surrounding cell and sensory hair cell types.


# Question 2: Assessing overall data quality
## Q2.1 Overall distributions
The expression values are currently in CPM. Here I will Log2 transform them so that the distribution is more evenly spread out and can be examined more easily.
```{r warning = FALSE}
logexpr <- column_to_rownames(expr, var = "gene") %>%
  log2() %>%
  rownames_to_column(var = "gene")

logexpr[1:5, 1:5] %>% kable()
```

I will examine the distribution of gene expression across all samples:  

### Boxplot
```{r}
# melt the data by sample
logmelt <- melt(logexpr, id.vars = "gene", variable.name = "sample", value.name = "expression")

# boxplot
ggplot(logmelt, aes(sample, expression)) +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("Distribution of gene expression across samples") +
  labs(y = "expression (CPM)")
```

### Overlapping density plots
```{r}
ggplot(logmelt, aes(x = expression, colour = sample)) +
  geom_density() +
  ggtitle("Distribution of gene expression across samples") +
  labs(x = "expression (CPM)")
```

### Which two samples stand out as different, in terms of the distribution of expression values, compared to the rest?

> Observation: Samples GSM1463879 and GSM1463880 have more genes with low expression values compared to the other samples.


## Q2.2 How do the samples correlate with one another?
I will examine the correlations between samples using a heatmap.
```{r message=FALSE, warning = FALSE, fig.width=8, fig.height=5}
library(pheatmap)

# correlation matrix comparing samples
correlation <- column_to_rownames(logexpr, var = "gene") %>%
  cor(method = "pearson")

# prep the annotation df: consider cell_type, organism_part, age, and batch
anno <- column_to_rownames(design_age, var = "sample") %>%
  select(cell_type, organism_part, age, batch) %>%
  as.data.frame()

pheatmap(mat = correlation, 
         cluster_rows = T, 
         cluster_cols = T, 
         annotation = anno,
         main = "Correlation between samples"
         )
```


### Among the factors cell_type, organism_part, age, and batch, which one seems to be most strongly correlated with clusters in gene expression data?

> The cell type is most strongly correlated with the clusters in gene expression data.

### There is a sample whose expression values correlate with the samples of the different cell_type just as well as with the samples of the same cell_type. Identify this sample by its ID.

> The sample GSM1463872 is a sensory hair cell sample, but its expression profile also correlates with the surrounding cells.


# Question 3: Using PCA to dig deeper
## Q3.1 Perform PCA to summarize the samples.
I will perform PCA on scaled and centered data for all genes.
First, I will scale and center the data:
```{r warning=FALSE}
sprDat <- column_to_rownames(expr, var = "gene")
sprDat <- t(scale(t(sprDat)))

round(data.frame(avgBefore = rowMeans(head(expr[,2:ncol(expr)])), avgAfter = rowMeans(head(sprDat)), 
    varBefore = apply(head(expr[,2:ncol(expr)]), 1, var), varAfter = apply(head(sprDat), 1, var)), 
    2) %>% head() %>% kable()
```

Second, I will perform PCA analysis:
```{r}
pcs <- prcomp(x = sprDat, scale. = F, center = F)
```

Here I will plot the variance explained by each PC:
```{r}
plot(pcs, main = "Variance explained by PCs", type = "l")
title(xlab = "PC")
```


##Q3.2 Confirm your suspicion. Is cell_type the biggest contributor to variation?
I will plot cell type on the x-axis and PC1, 2, or 3 on the y-axis, and make a visual assessment as to which PC is associated with cell type.
```{r warning = FALSE}
rotn <- pcs$rotation %>% as.data.frame() %>% rownames_to_column(var = "sample")
pcres <- left_join(design_age, rotn, by = "sample")

pc1_3 <- select(pcres, sample, cell_type, PC1:PC3) %>%
  melt(id.vars = c("sample", "cell_type"), variable.name = "PC", value.name = "rotation")

ggplot(pc1_3, aes(cell_type, rotation, colour = cell_type)) +
  geom_jitter(size = 2, width = 0.1) +
  facet_wrap(~PC) +
  theme_gray(base_size = 14) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")
```

> By visual assessment, it appears that PC1 is associated with cell_type. Therefore, cell_type is the biggest contributor to variation.

## Q3.3 Characterizing batch effects.
I will quantitatively assess the association of the batch variable with each PC up to PC10.

* __Definition__: The R-squared value (from a linear model) is interpreted as the proportion of the variance in the dependent variable that is predictable from the independent variable.
```{r warning = FALSE}
# select PC's 1 to 10
pc1_10 <- select(pcres, sample, batch, PC1:PC10) %>%
  gather(key = "PC", value = "rotation", PC1:PC10)

# make a numeric column for the PC number
pc1_10 <- mutate(pc1_10, PC_number = substr(pc1_10$PC, start = 3, stop = 4))

# make empty list to store results
rsquares <- list()

# loop to assess R-squared values
for(i in 1:10){
  subset <- filter(pc1_10, PC_number == i)
  model <- lm(formula = rotation ~ batch, data = subset) %>% summary()
  rsquares[i] <- model$r.squared
}


rsquares_df <- data.frame(PC = seq(1,10,1), r.squared = as.numeric(rsquares))

ggplot(rsquares_df, aes(PC, r.squared)) +
  geom_point() +
  geom_path() +
  scale_x_continuous(breaks = seq(1,10,1)) +
  theme_gray(base_size = 14) +
  ggtitle("Correlation of PC1-10 with Batch") +
  labs(y = "R-squared value")
```

#### How much of the variation in PC2 is explained by batch effects? How does that compare with PC1?

* `r signif(rsquares_df[rsquares_df$PC == 2,]$r.squared*100, 3)`% of the variation in PC2 is explained by batch effects.

* `r signif(rsquares_df[rsquares_df$PC == 1,]$r.squared*100, 3)`% of the variation in PC1 is explained by batch effects.

> Therefore, a greater proportion of the variation in PC2 is explained by batch effects compared to PC1.


